### PR TITLE
"Help, help, I'm being opressed" - the Oppressor upgrade module

### PR DESCRIPTION
**Pull request by InsaneHyena**
:cl: 
add: Added a roboticist-specific traitor item - the Oppressor upgrade module.
/:cl:

Adds an expensive traitor item that allows a roboticist to create a single security cyborg bound to do his evil bidding. The security cyborg comes in an obvious syndicate paintjob.

This PR was previously vetoed by PKPenguin, because he hates silicons and secborgs specifically. Now that his term is over, I'm posting an updated version.

### CODE REVIEW

**Comment by praisenarsie** [comment]
oh god

**Comment by InsaneHyena** [comment]
Whoops, my shitcode doesn't compile. Let's see how I fucked it up this time.

**Comment by InsaneHyena** [comment]
You saw nothing.

**Comment by DoctoreMad** [comment]
Okay lets pretend I didn't see that.
Your PR still doesn't merge.

**Comment by InsaneHyena** [comment]
I think I fixed everything this time.

**Comment by Repukan** [comment]
Sorry for asking but why the fuck did PK get a say on code things when he was only headmin?

**Comment by InsaneHyena** [comment]
Oranges asked for headmin opinion, because secborgs.

**Comment by Armhulen** [comment]
i am a headmin and this sounds like fun :+1:

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 1]
```diff
@@ -401,3 +401,18 @@
 
 	R.make_shell(src)
 	return TRUE
+
+/obj/item/borg/upgrade/traitorsec
+	name = "Bloodhound upgrade module"
+	desc = "Ilegal upgrade board that unlocks the banned security module."
+	icon_state = "cyborg_upgrade3"
+	require_module = 1
```
Comment (GunHog) [inline code/game/objects/items/robot/robot_upgrades.dm]: This simply makes the borg have to choose a module before installing the upgrade. That var should not be needed, since this upgrade is in of itself a module changer.

**Comment by InsaneHyena** [comment]
Also, crowdsourcing a more security or syndicate-related name than Bloodhound (which is originally just a name for a specific secborg sprite). I'm thinking Opressor sounds good.

**Comment by DoctoreMad** [comment]
If you make people scream SJW catchphrases when zapped by a Opressor's baton, I'm in.

**Comment by GunHog** [comment]
I was thinking "Infiltrator", or "Combat", but "Oppressor" sounds entirely fitting!

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 2]
```diff
@@ -403,7 +403,7 @@
 	return TRUE
 
 /obj/item/borg/upgrade/traitorsec
-	name = "Bloodhound upgrade module"
+	name = "Opressor upgrade module"
```
Comment (GunHog) [inline code/game/objects/items/robot/robot_upgrades.dm]: The correct spelling is "Oppressor". You need two "p"s.

**Comment by Armhulen** [comment]
I'd prefer suppressor but that works

File: code/game/objects/items/robot/robot_upgrades.dm (hunk 2 above)
Comment (InsaneHyena) [reply code/game/objects/items/robot/robot_upgrades.dm]: Don't lecture me on your Burgerstan language!

**Comment by factoryman942** [comment]
keep it with one p then you can call it an orange presser

File: code/modules/uplink/uplink_item.dm [hunk 3]
```diff
@@ -1313,6 +1313,13 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
 	limited_stock = 2 //you can't use more than two!
 	restricted_roles = list("Shaft Miner")
 
+/datum/uplink_item/role_restricted/traitorsec
+	name = "Oppressor Upgrade Module"
+	desc = "A box containing an cyborg upgrade module that instantly hacks a cyborg and unlocks the outlawed securiy module (in a inconspicuous black-and-red paintjob)"
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: securiy

Comment (ShizCalev) [reply code/modules/uplink/uplink_item.dm]: Also it's not explicitly outlawed, turning them off is a config setting. 

https://github.com/tgstation/tgstation/blob/6401bb8264708d3fc55c22fd6f7ba261aebc11ee/code/controllers/configuration.dm#L142
https://github.com/tgstation/tgstation/blob/7567fbafc43fa6430f165e6c702e5db5907d5733/code/modules/mob/living/silicon/robot/robot.dm#L212

File: code/modules/uplink/uplink_item.dm [hunk 4]
```diff
@@ -1313,6 +1313,13 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
 	limited_stock = 2 //you can't use more than two!
 	restricted_roles = list("Shaft Miner")
 
+/datum/uplink_item/role_restricted/traitorsec
+	name = "Oppressor Upgrade Module"
+	desc = "A box containing an cyborg upgrade module that instantly hacks a cyborg and unlocks the outlawed security module (in a inconspicuous black-and-red paintjob)"
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: The module isn't explicitly outlawed, turning them off is a config setting. It seems like a weird continuity break to have the description say such if someone reenables them on their server.

https://github.com/tgstation/tgstation/blob/6401bb8264708d3fc55c22fd6f7ba261aebc11ee/code/controllers/configuration.dm#L142
https://github.com/tgstation/tgstation/blob/7567fbafc43fa6430f165e6c702e5db5907d5733/code/modules/mob/living/silicon/robot/robot.dm#L212
https://github.com/tgstation/tgstation/blob/dab51af41ef82c5f817fb0530a65398bf8910738/config/game_options.txt#L262

Comment (GunHog) [reply code/modules/uplink/uplink_item.dm]: Perhaps use text such as "Unlocks a hacked version of the Security module" instead?

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 5]
```diff
@@ -401,3 +401,17 @@
 
 	R.make_shell(src)
 	return TRUE
+
+/obj/item/borg/upgrade/traitorsec
+	name = "Oppressor upgrade module"
+	desc = "Ilegal upgrade board that unlocks the banned security module."
```
Comment (ShizCalev) [inline code/game/objects/items/robot/robot_upgrades.dm]: Illegal*

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 6]
```diff
@@ -401,3 +401,17 @@
 
 	R.make_shell(src)
 	return TRUE
+
+/obj/item/borg/upgrade/traitorsec
+	name = "Oppressor upgrade module"
+	desc = "Illegal upgrade board that unlocks the banned security module."
```
Comment (ShizCalev) [inline code/game/objects/items/robot/robot_upgrades.dm]: Also same here, it's a config setting and not explicitly banned.

Should set the descriptions based on config.forbid_secborg being true or false imo

**Comment by InsaneHyena** [comment]
The Oppressor is literally only made because secborgs are turned off on /tg/ and are never coming back. It has no point otherwise, since it's a secborg but worse (obviously valid and expensive to make). If downstream servers don't have secborgs turned off, they don't need to merge this.

**Comment by DoctoreMad** [comment]
Bad: Just enabling this thing by default
Better(?): Purchasing this is enabled automatically by disabling secborgs in global config
Better(?): Purchasing this is enabled by separate config

**Comment by ExcessiveUseOfCobblestone** [comment]
just comment it out and stop needing to be handheld DOWNSTREAMERS!

**Comment by ShizCalev** [comment]
@insanehyena

That's a terrible justification for the current descriptions when you can just reword it to say "outlawed on some stations" or check the status of the secborg config to change it for you.

That said, to piggyback off of @doctremad, I would be in favor of this only being enabled if secborg are disabled. That would also resolve the description issues.

**Comment by ShizCalev** [comment]
You shouldn't have to comment out code to change config settings when you can just write your code to check what the config setting is set to in the first place.

**Comment by Bluespace** [comment]
Giving it an obvious sprite is going to get the shit kicked out of it 5 seconds in when John Assistant busts out his flash he stole from tech storage.

**Comment by InsaneHyena** [comment]
I would rather have it suck and buff it later, than make it overpowered and get it removed.

**Comment by DoctoreMad** [comment]
He has ranged tasers and tasers have a in-hand sprite.
If that's too much just follow your little exosuit so you know if you should kill yourself because you wasted your time implanting the module on a retarded player.

**Comment by DoctoreMad** [comment]
Shit. FLASHES have a in-hand sprite.

**Comment by Slignerd** [comment]
Isn't this basically #24628?

**Comment by GunHog** [comment]
@Slignerd 

> "This PR was previously vetoed by PKPenguin, because he hates silicons and secborgs specifically. Now that his term is over, I'm posting an updated version."

Posted in the OP

**Comment by Bawhoppen** [comment]
Why does everyone keep misspelling it?

**Comment by Slignerd** [comment]
I think Bloodhound was a better name.

**Comment by The-clown-known-as-Pepe** [comment]
@Slignerd if you actually read it, its black and red syndicate version

**Comment by PKPenguin321** [comment]
Secborgs are anti-fun regardless of what side they're fighting for, except for the secborg itself and its master

I have no holding on this and I already know you guys wont be convinced (at least not be me because "lol he's pk") but never say you weren't told in advance

**Comment by Repukan** [comment]
Lol it's PK

**Comment by kevinz000** [comment]
make me the first oppressor so i can remind everyone of why secborgs were disabled ez

File: code/modules/uplink/uplink_item.dm (hunk 4 above)
Comment (kevinz000) [reply code/modules/uplink/uplink_item.dm]: honestly if the config allows secborgs this module shouldn't be purchasable that's just going to be a waste of TC.

**Comment by Repukan** [comment]
So you can nerf flashes or emp nades after round end?

File: code/modules/uplink/uplink_item.dm (hunk 4 above)
Comment (ShizCalev) [reply code/modules/uplink/uplink_item.dm]: Add a new check & continue [here](https://github.com/InsaneHyena/tgstation/blob/8851267ad8a821363eee857e96b0fe065d9f8510/code/modules/uplink/uplink_item.dm#L30) & add two new list vars, `enabled_config` & `disabled_config` on the uplink_item parent so you can assign certain config values to disable/enable certain bundles, similar to how include_modes works, imo

Then have the var on the actual datum be something like `enabled_config = list(config.forbid_secborg)` or something like that.

**Comment by YakumoChen** [comment]
reminder that all it takes is one flashbang to end a borg

**Comment by Bluespace** [comment]
Flashbangs? Don't even need that, just grab a laser pointer to stun at range, move into for the kill with a flash and fire extinguisher.

**Comment by Durkel** [comment]
I'd rather not see this make it in but if it does it's going to need a min pop. Having a sec borg zoom around on a low pop shitting on everyone when there's not even a sec officer is just a recipe for butt hurt.

File: code/game/objects/items/storage/uplink_kits.dm [hunk 7]
```diff
@@ -305,3 +305,8 @@
 /obj/item/storage/box/syndie_kit/mimery/PopulateContents()
 	new /obj/item/spellbook/oneuse/mimery_blockade(src)
 	new /obj/item/spellbook/oneuse/mimery_guns(src)
+
+/obj/item/storage/box/syndie_kit/traitorsec_boards/New()
+	..()
+	contents = list()
```
Comment (Cyberboss) [inline code/game/objects/items/storage/uplink_kits.dm]: This should be PopulateContents() with no call to ..()

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Along with the secborg config issue...

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 8]
```diff
@@ -401,3 +401,17 @@
...
+	icon_state = "cyborg_upgrade3"
+	origin_tech = "data=4;combat=4;syndicate=3"
+
+/obj/item/borg/upgrade/traitorsec/action(mob/living/silicon/robot/R,mob/living/user)
+	if(..())
+		return 0
+
+	R.module.transform_to(/obj/item/robot_module/security/traitorsec)
+	R.emag_act(user)
+	return 1
```
Comment (Cyberboss) [inline code/game/objects/items/robot/robot_upgrades.dm]: TRUE

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 9]
```diff
@@ -401,3 +401,17 @@
...
+
+/obj/item/borg/upgrade/traitorsec
+	name = "Oppressor upgrade module"
+	desc = "Illegal upgrade board that unlocks the banned security module."
+	icon_state = "cyborg_upgrade3"
+	origin_tech = "data=4;combat=4;syndicate=3"
+
+/obj/item/borg/upgrade/traitorsec/action(mob/living/silicon/robot/R,mob/living/user)
+	if(..())
+		return 0
```
Comment (Cyberboss) [inline code/game/objects/items/robot/robot_upgrades.dm]: FALSE

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/uplink/uplink_item.dm (hunk 4 above)
Comment (InsaneHyena) [reply code/modules/uplink/uplink_item.dm]: Fine, but I'm not actually sure how to add the needed check.

**Comment by Bawhoppen** [comment]
Just make borgs slippable (overloads their movement sensors) and be damaged by water sprayed at them. Easy.

**Comment by DoctoreMad** [comment]
@Supermichael777

**Comment by YakumoChen** [comment]
>waaaah secborgs

>we need to Nerf Borgs even harder now

**Comment by YakumoChen** [comment]
Why don't we just force pk module on every Borg so that they will all be awful and nobody will want to play them

**Comment by YakumoChen** [comment]
Or make all Borgs into armless drones

**Comment by Slignerd** [comment]
Just remove silicons and borgs altogether, they're clearly not meant to be fun to play.

**Comment by InsaneHyena** [comment]
Thanks a lot to Antur for helping me with my shit code

File: code/modules/uplink/uplink_item.dm [hunk 10]
```diff
@@ -1313,6 +1318,17 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
...
+	item = /obj/item/storage/box/syndie_kit/traitorsec_boards
+	cost = 16
+	player_minimum = 20
+	restricted_roles = list("Roboticist", "Research Director")
+	
+/datum/uplink_item/role_restricted/traitorsec/SpecialRequirement()
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: Missing the s at the end.

Comment (InsaneHyena) [reply code/modules/uplink/uplink_item.dm]: Oh, so that's what I did wrong

Comment (GunHog) [reply code/modules/uplink/uplink_item.dm]: You should consider compiling this **before** you upload to Github! (And do not use the web editor for code changes that may not compile)

File: code/modules/uplink/uplink_item.dm [hunk 11]
```diff
@@ -1313,6 +1318,17 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
...
+/datum/uplink_item/role_restricted/traitorsec
+	name = "Oppressor Upgrade Module"
+	desc = "A box containing an cyborg upgrade module that instantly hacks a cyborg and unlocks the outlawed security module (in a inconspicuous black-and-red paintjob)"
+	item = /obj/item/storage/box/syndie_kit/traitorsec_boards
+	cost = 16
+	player_minimum = 20
+	restricted_roles = list("Roboticist", "Research Director")
+	
+/datum/uplink_item/role_restricted/traitorsec/SpecialRequirements()
+	return !config.forbid_secborg
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: Is the `!` supposed to be there? Shouldn't it be so that if the config is enabled (meaning secborgs are disabled normally), the oppressor module would be present?

File: code/modules/uplink/uplink_item.dm [hunk 12]
```diff
@@ -34,7 +34,9 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
 					continue
 			if(I.player_minimum && I.player_minimum > GLOB.joined_player_list.len)
 				continue
-
+			if(I.SpecialRequirements())
+				continue
+
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: Blank tabs here

Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: Blank tabs here

File: code/modules/uplink/uplink_item.dm [hunk 13]
```diff
@@ -1313,6 +1318,17 @@ GLOBAL_LIST_EMPTY(uplink_items) // Global list so we only initialize this once.
...
+	name = "Oppressor Upgrade Module"
+	desc = "A box containing an cyborg upgrade module that instantly hacks a cyborg and unlocks the outlawed security module (in a inconspicuous black-and-red paintjob)"
+	item = /obj/item/storage/box/syndie_kit/traitorsec_boards
+	cost = 16
+	player_minimum = 20
+	restricted_roles = list("Roboticist", "Research Director")
+	
+/datum/uplink_item/role_restricted/traitorsec/SpecialRequirements()
+	return config.forbid_secborg
+
```
Comment (ShizCalev) [inline code/modules/uplink/uplink_item.dm]: here too.

**Review by ShizCalev — COMMENTED** [review]
Tabs

**Comment by Anonymous583** [comment]
Just bring back Secborgs. Assistant feelings are irrelevant.

**Comment by optimumtact** [comment]
I'm chasing the headmins on this but assuming they're okay with this I can't see a reason to deny it.

**Comment by GunHog** [comment]
@Armhulen signed off on it earlier in this PR, so just wake Kor or the other guy.

**Comment by optimumtact** [comment]
please dont' tell me things that are already trivially obvious

**Comment by KorPhaeron** [comment]
If someone really wants to give away his traitor round to a robot who are we to stop them

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 14]
```diff
@@ -401,3 +401,17 @@
...
+
+/obj/item/borg/upgrade/traitorsec
+	name = "Oppressor upgrade module"
+	desc = "Illegal upgrade board that unlocks the banned security module."
+	icon_state = "cyborg_upgrade3"
+	origin_tech = "data=4;combat=4;syndicate=3"
+
+/obj/item/borg/upgrade/traitorsec/action(mob/living/silicon/robot/R,mob/living/user)
+	if(..())
+		return FALSE
```
Comment (optimumtact) [inline code/game/objects/items/robot/robot_upgrades.dm]: what is the purpose of this line

Comment (InsaneHyena) [reply code/game/objects/items/robot/robot_upgrades.dm]: I'm honestly not sure, Anturk told me to do it.

Comment (GunHog) [reply code/game/objects/items/robot/robot_upgrades.dm]: The parent of `action` contains the insertion checks (prevents use on dead borgs and enforces module restrictions on upgrades designed for a specific module). You can safely remove the `FALSE`, but you must keep the return. The attackby in `robot.dm` calls this to check if the upgrade was successfully installed.

**Comment by InsaneHyena** [comment]
Whoops, it broke shit.

**Comment by GunHog** [comment]
That is not why it broke, remove the `FALSE` again. It is related to `config.forbid_secborg`. 

Consider testing your work locally. You would find problems much faster if you bothered to at least compile first. You need to investigate why your config checker is broken. The code pointed out by Travis is here this "code/modules/uplink/uplink_item.dm:1330:error: returnconfig.forbid_secborg: undefined var". You can read the errors by clicking on the little red X by your commits.

**Comment by InsaneHyena** [comment]
>You need to merge your branch with current /tg/ master 

I need to do what now

**Comment by GunHog** [comment]
Okay...you might want to get on IRC so we can talk you through this.

**Comment by InsaneHyena** [comment]
Fuck's sake, I'll just reopen this PR later.

**Comment by GunHog** [comment]
:sob:

**Comment by InsaneHyena** [comment]
If you feel like fixing this shit right now, you can copy the code and open your own PR

**Comment by optimumtact** [comment]
saves me closing it lol
